### PR TITLE
Added bridging folder and HLO bridge

### PR DESCRIPTION
**Pull request by michael-isaev**
HLO bridge interface Paragraph with Tensorflow via XLA compiler and HLO.
Necessary dependencies on Tensorflow are added to WORKSPACE to be pulled
automatically.
hlo_bridge binary converts provided hlo file in .pb or .textproto
formats and creates a paragraph file with respect to the provided
options.

### CODE REVIEW

File: paragraph/bridging/hlo/BUILD [hunk 1]
```diff
@@ -0,0 +1,70 @@
+load("@org_tensorflow//tensorflow:tensorflow.bzl", "tf_cc_binary")
```
Comment (nicmcd) [inline paragraph/bridging/hlo/BUILD]: This isn't being used

File: paragraph/bridging/hlo/hlo_bridge.cc [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+#include "absl/flags/flag.h"
+#include "absl/flags/parse.h"
+#include "paragraph/bridging/hlo/hlo_converter.h"
+#include "paragraph/graph/graph.h"
+#include "tensorflow/compiler/xla/tools/hlo_module_loader.h"
+
+ABSL_FLAG(std::string, hlo_module, "",
+          "A path to a file containing the HLO module.");
+ABSL_FLAG(std::string, target_dir, "",
```
Comment (nicmcd) [inline paragraph/bridging/hlo/hlo_bridge.cc]: Why can't you just have a single output flag?

File: paragraph/bridging/hlo/hlo_bridge.cc [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+  };
+
+  std::unique_ptr<xla::HloModule> module;
+  if (!module_path.empty()) {
+    xla::hlo_module_loader_details::Config loader_config;
+    loader_config.num_replicas = num_replicas;
+    auto module_statusor = xla::LoadModuleFromFile(module_path, loader_config);
+    TF_CHECK_OK(module_statusor.status());
+    module = std::move(module_statusor.ValueOrDie());
+  }
```
Comment (nicmcd) [inline paragraph/bridging/hlo/hlo_bridge.cc]: You are allowing empty input modules?

File: paragraph/bridging/hlo/hlo_bridge.cc [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+    filename = target_name;
+  }
+
+  std::string file_path = absl::StrCat(target_path,
+                                       "/",
+                                       filename,
+                                       ".paragraph",
+                                       target_extension);
+  CHECK_OK(graph->WriteToFile(file_path));
+}
```
Comment (nicmcd) [inline paragraph/bridging/hlo/hlo_bridge.cc:85]: return 0?

File: paragraph/bridging/hlo/hlo_converter.cc [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+  } else if (hlo_opcode == xla::HloOpcode::kOutfeed) {
+    return "outfeed";
+  } else {
+    return "delay";
+  }
+  return "delay";
+}
+
+xla::HloCostAnalysis::Properties GetPropertiesForInstruction(
+    const int64_t id,
```
Comment (nicmcd) [inline paragraph/bridging/hlo/hlo_converter.cc]: There is no reason to use const for types that are passed by value. There are a couple of these in here.

File: paragraph/bridging/hlo/hlo_converter.cc [hunk 6]
```diff
@@ -0,0 +1,300 @@
...
+        const ComputeCostAnalysis::Properties per_second_rates) {
+  auto cost_analysis = absl::make_unique<ComputeCostAnalysis>(
+      ShapeSize, per_second_rates);
+  xla::HloComputation* hlo_computation = module->entry_computation();
+  TF_CHECK_OK(hlo_computation->Accept(cost_analysis.get()));
+  TF_CHECK_OK(cost_analysis->UpdateInstructionProperties());
+  return cost_analysis;
+}
+
+xla::StatusOr<paragraph::GraphProto> HloConverter(
```
Comment (nicmcd) [inline paragraph/bridging/hlo/hlo_converter.cc:257]: This is the only public function declared in the header file. Put all others in an anonymous namespace.

File: paragraph/bridging/hlo/BUILD (hunk 1 above)
Comment (michael-isaev) [reply paragraph/bridging/hlo/BUILD]: done

File: paragraph/bridging/hlo/hlo_bridge.cc (hunk 2 above)
Comment (michael-isaev) [reply paragraph/bridging/hlo/hlo_bridge.cc]: Done

File: paragraph/bridging/hlo/hlo_bridge.cc (hunk 3 above)
Comment (michael-isaev) [reply paragraph/bridging/hlo/hlo_bridge.cc]: Fixed

File: paragraph/bridging/hlo/hlo_bridge.cc (hunk 4 above)
Comment (michael-isaev) [reply paragraph/bridging/hlo/hlo_bridge.cc:85]: Done

File: paragraph/bridging/hlo/hlo_converter.cc (hunk 6 above)
Comment (michael-isaev) [reply paragraph/bridging/hlo/hlo_converter.cc:257]: Done

File: paragraph/bridging/hlo/hlo_converter.cc (hunk 5 above)
Comment (michael-isaev) [reply paragraph/bridging/hlo/hlo_converter.cc]: Fixed

**Pull request merged**